### PR TITLE
feat: add ifs_same_cond

### PR DESCRIPTION
**Pull request by Gianfranco99**
Summary 📌
This PR Resolves: https://github.com/keep-starknet-strange/cairo-lint/issues/87

What it does:
This PR introduces a new lint that detects consecutive if statements with the same condition. When two if or else if blocks check the same condition, this could be an indication of a copy & paste error, and this lint helps prevent such mistakes.
All tests pass
image

### CODE REVIEW

File: crates/cairo-lint-core/src/lints/ifs/ifs_same_cond.rs [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+fn is_function_call(expr: &Expr) -> bool {
+    matches!(expr, Expr::FunctionCall(_))
+}
+
+fn get_condition_text(db: &dyn SyntaxGroup, condition: &Condition) -> Option<String> {
+    match condition {
+        Condition::Expr(expr) => Some(expr.as_syntax_node().get_text(db).to_string()),
+        Condition::Let(condition_let) => Some(condition_let.expr(db).as_syntax_node().get_text(db).to_string()),
+    }
+}
```
Comment (0xLucqs) [inline crates/cairo-lint-core/src/lints/ifs/ifs_same_cond.rs:18]: why does it return an option ?

File: crates/cairo-lint-core/src/lints/ifs/ifs_same_cond.rs [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+use cairo_lang_diagnostics::Severity;
+use cairo_lang_syntax::node::ast::{BlockOrIf, Condition, Expr, ExprIf, OptionElseClause};
+use cairo_lang_syntax::node::db::SyntaxGroup;
+use cairo_lang_syntax::node::TypedSyntaxNode;
+
+pub const DUPLICATE_IF_CONDITION: &str = "Consecutive `if` with the same condition found.";
+
+fn is_function_call(expr: &Expr) -> bool {
+    matches!(expr, Expr::FunctionCall(_))
+}
```
Comment (0xLucqs) [inline crates/cairo-lint-core/src/lints/ifs/ifs_same_cond.rs:11]: i don't think this is necessary you can directly use what's inside the function

File: crates/cairo-lint-core/src/lints/ifs/ifs_same_cond.rs [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+fn get_condition_text(db: &dyn SyntaxGroup, condition: &Condition) -> Option<String> {
+    match condition {
+        Condition::Expr(expr) => Some(expr.as_syntax_node().get_text(db).to_string()),
+        Condition::Let(condition_let) => Some(condition_let.expr(db).as_syntax_node().get_text(db).to_string()),
+    }
+}
+
+pub fn check_duplicate_if_condition(db: &dyn SyntaxGroup, if_expr: &ExprIf, diagnostics: &mut Vec<PluginDiagnostic>) {
```
Comment (0xLucqs) [inline crates/cairo-lint-core/src/lints/ifs/ifs_same_cond.rs:20]: you can have multiple `else if` the code doesn't seem to handle that case

File: crates/cairo-lint-core/tests/test_files/ifs/ifs_same_cond [hunk 4]
```diff
@@ -0,0 +1,167 @@
+//! > Same condition with boolean
```
Comment (0xLucqs) [inline crates/cairo-lint-core/tests/test_files/ifs/ifs_same_cond:1]: add a tests with:
- more than one else if
- one where 2 conditions are similar but not the last one
and try to think about more edge cases to avoid bugs

**Review by 0xLucqs — CHANGES_REQUESTED** [review]

File: crates/cairo-lint-core/src/lints/ifs/ifs_same_cond.rs (hunk 2 above)
Comment (Gianfranco99) [reply crates/cairo-lint-core/src/lints/ifs/ifs_same_cond.rs:11]: Do you mean using it like this 
```
      let is_if_function_call = match &if_condition {
                Condition::Expr(expr) =>  matches!(&expr.expr(db), Expr::FunctionCall(_)),
                Condition::Let(condition_let) => matches!(&condition_let.expr(db),Expr::FunctionCall(_)),
            };

            let is_else_if_function_call = match &else_if_condition {
                Condition::Expr(expr) => matches!(&expr.expr(db), Expr::FunctionCall(_)),
                Condition::Let(condition_let) => matches!(&condition_let.expr(db),Expr::FunctionCall(_)),
            };

            if is_if_function_call || is_else_if_function_call {
                return;
            }
```

Comment (Gianfranco99) [reply crates/cairo-lint-core/src/lints/ifs/ifs_same_cond.rs:11]: or do you like it better like this?
```
if matches!(&if_condition, Condition::Expr(expr) if matches!(expr.expr(db), Expr::FunctionCall(_)))
    || matches!(&if_condition, Condition::Let(condition_let) if matches!(condition_let.expr(db), Expr::FunctionCall(_)))
{
    return; // Salimos si la condición es una llamada a función
}
```

Comment (Gianfranco99) [reply crates/cairo-lint-core/src/lints/ifs/ifs_same_cond.rs:11]: Now that I think about it, I think we can only validate with the if if it is a function or not, right? because if the if is not a function but the else_if is, they would no longer match and it would not apply to the case

File: crates/cairo-lint-core/tests/test_files/ifs/ifs_same_cond [hunk 5]
```diff
@@ -0,0 +1,370 @@
+//! > Combined conditions with different if
```
Comment (0xLucqs) [inline crates/cairo-lint-core/tests/test_files/ifs/ifs_same_cond:1]: what if you use literals in the if condition ? `if 1 == 1` `if a == 3`

**Review by 0xLucqs — CHANGES_REQUESTED** [review]
add these tests otherwise looks good

**Comment by 0xLucqs** [comment]
any news ?

**Comment by Gianfranco99** [comment]
> any news ?
 
Done,  I'm going to do a push now. @0xLucqs

File: crates/cairo-lint-core/src/lints/ifs/ifs_same_cond.rs [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+        }
+    }
+}
+
+fn are_conditions_equivalent(condition1: &str, condition2: &str) -> bool {
+    condition1.contains("==")
+        && condition2.contains("==")
+        && condition1.split("==").nth(0).unwrap().trim() == condition2.split("==").nth(1).unwrap().trim()
+        && condition1.split("==").nth(1).unwrap().trim() == condition2.split("==").nth(0).unwrap().trim()
+}
```
Comment (0xLucqs) [inline crates/cairo-lint-core/src/lints/ifs/ifs_same_cond.rs:88]: What are you trying to achieve here ?

File: crates/cairo-lint-core/src/lints/ifs/ifs_same_cond.rs [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+                    BinaryOperator::Eq(_) => {
+                        return format!("{} == {}", lhs_text, rhs_text);
+                    }
+                    BinaryOperator::LT(_) => {
+                        return format!("({}, {})", rhs_text, lhs_text);
+                    }
+                    BinaryOperator::GT(_) => {
+                        return format!("({}, {})", lhs_text, rhs_text);
+                    }
+                    _ => {}
```
Comment (0xLucqs) [inline crates/cairo-lint-core/src/lints/ifs/ifs_same_cond.rs:71]: what is that ?

**Review by 0xLucqs — CHANGES_REQUESTED** [review]
still need the `1==1` test

File: crates/cairo-lint-core/src/lints/ifs/ifs_same_cond.rs (hunk 6 above)
Comment (Gianfranco99) [reply crates/cairo-lint-core/src/lints/ifs/ifs_same_cond.rs:88]: What I do here is validate if the conditions of the if and the else if are the same even if they are changed in the order in the else if
example => `fn main(){
    let a = 3;

    if a == 3 {
        println!("a == 3");
    } else if 3 == a {
        println!("3 == a");
    }
}`

File: crates/cairo-lint-core/src/lints/ifs/ifs_same_cond.rs (hunk 7 above)
Comment (Gianfranco99) [reply crates/cairo-lint-core/src/lints/ifs/ifs_same_cond.rs:71]: this is to match the conditions = , < , > 
example =>
fn main(){
    let a:u32 = 3;

    if a > 3 {
        println!("a == 3");
    } else if 3 < a {
        println!("3 == a");
    }
}

File: crates/cairo-lint-core/src/lints/ifs/ifs_same_cond.rs [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+            let else_if_condition = else_if_block.condition(db);
+
+            if matches!(&else_if_condition, Condition::Expr(expr) if matches!(expr.expr(db), Expr::FunctionCall(_)))
+                || matches!(&else_if_condition, Condition::Let(condition_let) if matches!(condition_let.expr(db), Expr::FunctionCall(_)))
+            {
+                current_block = else_if_block.else_clause(db);
+                continue;
+            }
+
+            let if_condition_text = get_condition_text(db, &if_condition);
```
[the remark this answers is not in the document]
Comment (Gianfranco99) [reply crates/cairo-lint-core/src/lints/ifs/ifs_same_cond.rs:31]: You're right, I can leave it out of the while.

**Comment by Gianfranco99** [comment]
> still need the `1==1` test

I had added that test but I deleted it since there was a conflict with a new test that validates the same condition and it generated conflicts in the file showing another error message, but if you want we can see how we solve this and also add that test

File: crates/cairo-lint-core/src/lints/ifs/ifs_same_cond.rs (hunk 7 above)
Comment (0xLucqs) [reply crates/cairo-lint-core/src/lints/ifs/ifs_same_cond.rs:71]: what do you do with the output, for example the function returns `(1, a)` how is that helpful ?

File: crates/cairo-lint-core/src/lints/ifs/ifs_same_cond.rs (hunk 6 above)
Comment (0xLucqs) [reply crates/cairo-lint-core/src/lints/ifs/ifs_same_cond.rs:88]: i mean you can just put the variables in a hashset and check its length ?

**Review by 0xLucqs — APPROVED** [review]

**Comment by 0xLucqs** [comment]
@all-contributors please add @Gianfranco99 for code and tests
